### ISSUE TITLE
Cloudflare DDOS protection

### ISSUE THREAD

**Issue by guryaniv** [ultrafunkamsterdam/undetected-chromedriver#46]
Hi,

working with the latest version (1.5) , with ChromeDriver 85.0.4183.87,
I'm getting blocked by sites with cloudflare DDOS protection (Under Attack Mode).
Example of such site - [http://void.to](http://void.to)

driver gets detected and gets stuck in the "Checking your browser" page

**Comment by Dellgenius** [comment]
I am having the same issue. Have you found a solution yet by any chance?

**Comment by ultrafunkamsterdam** [comment]
Yes, after 200h+ hours I found the holy grail which is going to bypass CF as well as all small startup's . However a lot has to be rewritten which is what I'm working on.On Nov 8, 2020 01:38, tuffstuff <notifications@github.com> wrote:
I am having the same issue. Have you found a solution yet by any chance?

**Comment by chwba** [comment]
https://www.mc-market.org same issue

**Comment by Dellgenius** [comment]
That's great to hear! Would you mind sharing what it was that was causing the issue?

**Comment by rocbr** [comment]
I have the same issue !

**Comment by Avnsx** [comment]
Same issue too ):

**Comment by czoins** [comment]
Simply replace the "undetectedchromeDRiver" with something with equivalent length. Worked for me.

https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/7400dd407fa7d85559f580e46178a847831931dd/undetected_chromedriver/__init__.py#L233

**Comment by Avnsx** [comment]
> Simply replace the "undetectedchromeDRiver" with something with equivalent length. Worked for me.
> 
> https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/7400dd407fa7d85559f580e46178a847831931dd/undetected_chromedriver/__init__.py#L233

You're fucking genius, thanks man ✌

**Comment by ultrafunkamsterdam** [comment]
> Simply replace the "undetectedchromeDRiver" with something with equivalent length. Worked for me.
> 
> https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/7400dd407fa7d85559f580e46178a847831931dd/undetected_chromedriver/__init__.py#L233

Which would mean Cloudflare is checking my little repo for their antibot algo, that's pretty genius as well.
I assume you deleted your chromedriver executable after making the change so it patches it with your random input? 
That would be a simple change to implement if that really works, however i've tested this is as well, and did not have positive results. please reply if it's working for any of you.

**Comment by czoins** [comment]
> > Simply replace the "undetectedchromeDRiver" with something with equivalent length. Worked for me.
> > https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/7400dd407fa7d85559f580e46178a847831931dd/undetected_chromedriver/__init__.py#L233
> 
> Which would mean Cloudflare is checking my little repo for their antibot algo, that's pretty genius as well.
> I assume you deleted your chromedriver executable after making the change so it patches it with your random input?
> That would be a simple change to implement if that really works, however i've tested this is as well, and did not have positive results. please reply if it's working for any of you.

I have edited the script so it randomizes the string each time it executes. So yes, in another words  i have bascially deleted the chromdriver. I can confirm that it works, as when i placed back the old string it no longer does work. Whether this works for all cloudflare protected sites isn't clear, but don't they all use the same algorithm?

**Comment by guryaniv** [comment]
@ultrafunkamsterdam it's working for me. thanks @czoins

**Comment by frzsombor** [comment]
Related to #38 and means that it either took CF more than a month, to update their code to detect this, or they just found this repo and joined the "watch party" recently :)

**Comment by Avnsx** [comment]
> Related to #38 and means that it either took CF more than a month, to update their code to detect this, or they just found this repo and joined the "watch party" recently :)

he predicted this happening lmao 😳

**Comment by ultrafunkamsterdam** [comment]
lol. 
Seemed the issue was only in the first 3 characters being the same. after changing to xxz it also works as well.
I updated the package to 1.5.1 now.

Note: i did not implement the random function proposed and PR'd by @czoins (yet). And reason for that is different than most of you would think, however please let me know the minute this is issue returns.

Meanwhile In the background i'm still busy rewriting and testing a new monstrous function on all platforms (without prying eyes), which will be able to bypass even future enhancements. Thanks @czoins for your work.

**Comment by Krylanc3lo** [comment]
I have the same issue with tirexo.cc. 

@ultrafunkamsterdam, Let me know if I can assist you for the rewite by testing or otherwise

**Comment by chwba** [comment]
Cloudflare seems to detect it again, even with randomized 'undetectedchromeDRiver'. You can test on https://v3rmillion.net, https://www.sythe.org, mc_market

**Comment by chemeng** [comment]
Same here, cloudflare detection started 2 days ago for me.

**Comment by jjstecchino** [comment]
Same here as well. Is there any workaround to the new Cloudflare detection?

**Comment by azhuge233** [comment]
If it helps, you can use Playwright  ([Python](https://github.com/microsoft/playwright-python) [C#](https://github.com/microsoft/playwright-sharp)) instead, and choose Firefox or WebKit browser (Chromium somehow can still be detected).
It's similar to selenium and it can bypass Cloudflare protection even in headless mode, I've tried python and C# codes on SteamDB (which also uses CF protection), it works.

**Comment by khiemnd5** [comment]
Cloudflare still detect checking :(

**Comment by bl4ckOut** [comment]
> 
> 
> If it helps, you can use Playwright ([Python](https://github.com/microsoft/playwright-python) [C#](https://github.com/microsoft/playwright-sharp)) instead, and choose Firefox or WebKit browser (Chromium somehow can still be detected).
> It's similar to selenium and it can bypass Cloudflare protection even in headless mode, I've tried python and C# codes on SteamDB (which also uses CF protection), it works.

Works great. 
If you want to continue working with selenium you can just grab the cookies fom playwight context after visiting the site and add them to the driver instance.

**Comment by JAManfredi** [comment]
Somewhat of an ignorant question, but I assume youd want to grab the cookies if you're working with a form of authenticated session? Any other reasoning or benefits to that just as it pertains to avoiding bot detection?

**Comment by bl4ckOut** [comment]
I just want to get past the cloudflare protection. After you successfully loaded a site with playwright the corresponding cloudflare cookies (like 'cf_clearance' and '__cfduid') have been authenticated and can be used in other sessions. There are no other advantages. I just like selenium and don't want to learn a new scraper.

**Comment by JAManfredi** [comment]
Ah makes sense. Thanks for that, I wasnt aware that cloudflare stores those cookies.

**Comment by ultrafunkamsterdam** [comment]
I did not create this project to forward poeple to other repo's. Please stop "plugging" these solutions. It is not a solution to copy a cookie between browsers using 2 libraries of jou could also just solve the damn captcha in 2 seconds and be rid of it. I just finished draft, and will upload a video demonstrating it goes to several heavy protected hcaptcha/cloudflare/custom impl. sites. Give me another 2 days and the package will be updated.

**Comment by czoins** [comment]
As a workaround until the issue is fixed, you can automate chrome itself which isn't detected. I am not sure if you can run multiple instances though.

```python
import subprocess
#launch chrome with remote debug, and open your url (you can also use a separate user directory (--user-data-dir) which is advisable)
subprocess.Popen(['C:/Program Files (x86)/Google/Chrome/Application/chrome.exe', '--remote-debugging-port=7102', 'url'])
#might want to add a time.sleep here until chrome launches
chrome_options.add_experimental_option("debuggerAddress","localhost:7102")
driver = uc.Chrome(options=chrome_options)
#connected, use the same way as chromedriver
...
```

**Comment by ultrafunkamsterdam** [comment]
@czoins although i like the approach ( caugh ... it's how new version is implemented), and it works without patching as well, it **does** get detected by cf IUAM mode for example. Do you, by chance, have a working setup in headless mode. I'm running against a huge lock and memory builds FAST until 30G is used so i wonder if you got it working for headless.

**Comment by JAManfredi** [comment]
I know this is likely outside the scope of the library given its more about how it’s used, but at least 1 notable scraping repo seems to enforce a wait after a page loads when it detects the host is using cloudflare. I’m not sure how that could be incorporated here, but is anyone familiar with what detection that’s attempting to circumvent? Maybe there _something_ that could help.

**Comment by ultrafunkamsterdam** [comment]
exactly that. and using context managers.

**Comment by zpzyyzp** [comment]
> If it helps, you can use Playwright ([Python](https://github.com/microsoft/playwright-python) [C#](https://github.com/microsoft/playwright-sharp)) instead, and choose Firefox or WebKit browser (Chromium somehow can still be detected).
> It's similar to selenium and it can bypass Cloudflare protection even in headless mode, I've tried python and C# codes on SteamDB (which also uses CF protection), it works.

Playwright works as a charm.
Goodbye selenium.

**Comment by ultrafunkamsterdam** [comment]
to test:

`pip install -U undetected_chromedriver`

```python
import undetected_chromedriver.v2 as uc

driver = uc.Chrome()
with driver:
    driver.get('https://lhcdn.botprotect.io')
```

**Comment by jjstecchino** [comment]
I updated through pip as per your instructions but importing undetected_chromedriver.v2 fails

**Comment by ultrafunkamsterdam** [comment]
@jjstecchino,  pip does not update to beta versions. I did not know that, so for simplicity i released it, and now it should work.

**Comment by jjstecchino** [comment]
Thanks for doing that. I was able to get it. Now on executing

driver = uc.Chrome()

I get an error:

Exception has occurred: TypeError
expected str, bytes or os.PathLike object, not NoneType

I am on a Mac. chromedriver gets downloaded in my test project directory.

**Comment by jjstecchino** [comment]
trying to debug, it seems in v2.py find_chrome_executable() can't find chrome executable on osx (using Big Sur)

**Comment by MjedAl** [comment]
Still getting blocked by Cloudflare in v2 :(

**Comment by jjstecchino** [comment]
> Still getting blocked by Cloudflare in v2 :(

I fixed a few things on the v2 driver to find chrome executable on osx, but I can now go past Cloudflare

**Comment by ultrafunkamsterdam** [comment]
@MjedAl 

please try and close every chrome or chromedriver process, before starting.


```python

import undetected_chromedriver.v2 as uc
driver = uc.Chrome()

with driver:
    driver.get("https://coinfaucet.eu")
time.sleep(4) # sleep only used for timing of screenshot
driver.save_screenshot("coinfaucet.eu.png")

with driver:
    driver.get("https://cia.gov")
time.sleep(4) # sleep only used for timing of screenshot
driver.save_screenshot("cia.gov.png")

with driver:
    driver.get("https://lhcdn.botprotect.io")
time.sleep(4) # sleep only used for timing of screenshot
driver.save_screenshot("notprotect.io.png")

with driver:
    driver.get("https://www.datadome.co")
time.sleep(4) # sleep only used for timing of screenshot
driver.save_screenshot("datadome.co.png")
```

![coinfaucet eu](https://user-images.githubusercontent.com/21027969/104851615-70a51c00-58f6-11eb-86c3-faa35a6566ba.png)
![datadome co](https://user-images.githubusercontent.com/21027969/104851617-713db280-58f6-11eb-8ccb-d8d2c434b304.png)
![notprotect io](https://user-images.githubusercontent.com/21027969/104851619-71d64900-58f6-11eb-9e11-b2c660f783a2.png)
![cia gov](https://user-images.githubusercontent.com/21027969/104851620-71d64900-58f6-11eb-95ef-010926cc87d9.png)

@jjstecchino , bummer, i tested it on Catalina, which worked, i'll fix that asap.

**Comment by chwba** [comment]
Still doesnt go through - https://www.mc-market.org/

**Comment by kforbro** [comment]
> @MjedAl
> 
> please try and close every chrome or chromedriver process, before starting.
> 
> ```python
> import undetected_chromedriver.v2 as uc
> driver = uc.Chrome()
> 
> with driver:
>     driver.get("https://coinfaucet.eu")
> time.sleep(4) # sleep only used for timing of screenshot
> driver.save_screenshot("coinfaucet.eu.png")
> 
> with driver:
>     driver.get("https://cia.gov")
> time.sleep(4) # sleep only used for timing of screenshot
> driver.save_screenshot("cia.gov.png")
> 
> with driver:
>     driver.get("https://lhcdn.botprotect.io")
> time.sleep(4) # sleep only used for timing of screenshot
> driver.save_screenshot("notprotect.io.png")
> 
> with driver:
>     driver.get("https://www.datadome.co")
> time.sleep(4) # sleep only used for timing of screenshot
> driver.save_screenshot("datadome.co.png")
> ```
> 
> ![coinfaucet eu](https://user-images.githubusercontent.com/21027969/104852024-a6e39b00-58f8-11eb-83d3-50f5cc246b27.png)
> ![cia gov](https://user-images.githubusercontent.com/21027969/104852028-a814c800-58f8-11eb-8326-ab26e05e1fd3.png)
> ![notprotect io](https://user-images.githubusercontent.com/21027969/104852026-a814c800-58f8-11eb-9236-76da40c8be13.png)
> ![datadome co](https://user-images.githubusercontent.com/21027969/104852025-a814c800-58f8-11eb-8b41-0b41dc7e3b16.png)
> 
> @jjstecchino , bummer, i tested it on Catalina, which worked, i'll fix that asap.

Because of the import .v2 deletes all files in the project directory
https://youtu.be/8o02waJqd1k

**Comment by chemeng** [comment]
@ultrafunkamsterdam any ideas why/how CF detects chromedriver and why with remote debug it doesn't?

**Comment by angustohrallegrinski** [comment]
Hi! How can use the patch when using a selenium standalone (or hub) docker image, so using the Remote webdriver?

**Comment by ultrafunkamsterdam** [comment]
> @ultrafunkamsterdam any ideas why/how CF detects chromedriver and why with remote debug it doesn't?

@chemeng , likely because chrome is started within the chromedriver process (which makes it a direct fork/child of chromedriver), thus sharing environments, and using remote debugging you start both chrome and chromedriver  as seperate processes under your user. This also happens to enable the functionality to disconnect and reconnect without closing the browser window. This interferes the CF check which uses the samesite=LAX cookie, https://support.cloudflare.com/hc/en-us/articles/360038470312-Understanding-SameSite-cookie-interaction-with-Cloudflare )

**Comment by montovaneli** [comment]
> > @MjedAl
> > please try and close every chrome or chromedriver process, before starting.
> > ```python
> > import undetected_chromedriver.v2 as uc
> > driver = uc.Chrome()
> > 
> > with driver:
> >     driver.get("https://coinfaucet.eu")
> > time.sleep(4) # sleep only used for timing of screenshot
> > driver.save_screenshot("coinfaucet.eu.png")
> > 
> > with driver:
> >     driver.get("https://cia.gov")
> > time.sleep(4) # sleep only used for timing of screenshot
> > driver.save_screenshot("cia.gov.png")
> > 
> > with driver:
> >     driver.get("https://lhcdn.botprotect.io")
> > time.sleep(4) # sleep only used for timing of screenshot
> > driver.save_screenshot("notprotect.io.png")
> > 
> > with driver:
> >     driver.get("https://www.datadome.co")
> > time.sleep(4) # sleep only used for timing of screenshot
> > driver.save_screenshot("datadome.co.png")
> > ```
> > 
> > 
> > @jjstecchino , bummer, i tested it on Catalina, which worked, i'll fix that asap.
> 
> Because of the import .v2 deletes all files in the project directory
> https://youtu.be/8o02waJqd1k

Same here.

**Comment by RabbitHol3** [comment]
having the same problem, anyone found a solution ?

**Comment by ultrafunkamsterdam** [comment]
Closed. Open a new issue with code and error log if you think you are having this issue and you are 100% sure your ip is clean.

**Comment by udiudiudiudi** [comment]
> Simply replace the "undetectedchromeDRiver" with something with equivalent length. Worked for me.
> 
> https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/7400dd407fa7d85559f580e46178a847831931dd/undetected_chromedriver/__init__.py#L233

Can you please check if you find also a solution for trainline.com, apparently after 50 rounds of chromedriver, they still somehow manage to profile the driver and block it, with stormproxies

**Comment by rnyPlanet** [comment]
> If it helps, you can use Playwright ([Python](https://github.com/microsoft/playwright-python) [C#](https://github.com/microsoft/playwright-sharp)) instead, and choose Firefox or WebKit browser (Chromium somehow can still be detected).
> It's similar to selenium and it can bypass Cloudflare protection even in headless mode, I've tried python and C# codes on SteamDB (which also uses CF protection), it works.

is it working at the moment? i tried it and it is as detectable as chrome
